### DISCUSSION TITLE
Plan to make CameraView more stable

### DISCUSSION THREAD

**Discussion by pictos** [CommunityToolkit/Maui#206] [New Feature Discussions]
### Motivation

The `CameraView` control on XCT is far away from what we want as a stable and reliable control, right now we support 3 platforms Android, iOS, and UWP. From these 3 platforms, the Android implementation is a nightmare to maintain. In a general overview I can list those issues:

- Complex code base to maintain;
- Open bugs that we don't know how to fix (like the rotation image);
- Complex code to implement new features.

Is very terrifying to do any change on that codebase without breaking something that's working already. From that, I came with this idea/plan.

### The Plan

Since we're moving all controls to a new place this is the right moment to _break hard_, in other words, we should drop some features on our control and implement a very basic `CameraView`, make sure that's stable and solid and then do enhancements.

I know that's not ideal at first hand to drop features and break our users but if we have good community support on this we can be at the same level, in terms of features, that we're in XCT on the release date.

I believe that we can do the following steps to execute this plan.

## First Step

Agree on what features we will have for the first version of the `CameraView`. I would say that we should be able to take a picture and return the `Stream` or `byte[]` to the user.

## Second Step

We need to reduce the complexity of our codebase, as I said before the Android implement is the most complex one, to fix that we should get rid of the `Camera2` API and use the `CameraX` API, that's much easier to work with.

## Third Step

Implement missing features that we already have implemented on the XCT version. We don't need to implement all of them if we see that isn't too much value in adding the feature here (e.g. a feature that almost no one is using).

## Fourth Step

Implement new features and make everyone happy!

**Comment by danielftz** [comment]
I would like to give this a try. I previously implemented a Barcode Scanner (custom renderer + Camera X) in Xamarin that returned byte[] of YUVLuminanceSource from preview for further processing in shared project. If you would like, I can put together a demo project of the camera running just the preview, and see how it turns out. 

I am also wondering how the Handler architecture would factor in the porting of CameraView, since my implementation still uses the custom renderer architecture.

**Reply by pictos** [reply]
@danielftz if you have time feel free to put the parts together and share a demo. I still need to play with CameraX and see how architect everything, so any sample project is welcome :D

**Reply by danielftz** [reply]
https://github.com/danielftz/MCT-CameraView-Demo
Hi, have made a simple camera preview with Android CameraX.

**Reply by pictos** [reply]
Thanks @danielftz I'll take a look it

**Reply by danielftz** [reply]
I have updated the camera view project. Currently it will display a camera preview on Android and iOS. It also uses the handler architecture.

**Reply by pictos** [reply]
@danielftz thanks for the help and input. FYI this is in WIP mode now. I finished the Android part and need to circle back to finish the iOS and Windows (if the WinUI releases a CameraView support)

**Reply by danielftz** [reply]
> @danielftz thanks for the help and input. FYI this is in WIP mode now. I finished the Android part and need to circle back to finish the iOS and Windows (if the WinUI releases a CameraView support)

Awesome! Im also thinking for Barcode scanning, we could use MLKit for Android and VisionFramework for iOS.

**Reply by pictos** [reply]
@danielftz well, for that we have the ZXing.mobile... But we will not support this feature on MCT (maybe later, but not in a short time frame). With that in mind I create the manager in a way that you can easily apply some customizations to the CameraX (android) in order to get the image recognition features. So in other words, with our cameraView should be very easy to add/change the camera behavior

**Comment by xDaijobu** [comment]
Hey, please check this one too XD 

https://github.com/xDaijobu/ToToolkit

**Reply by pictos** [reply]
Thanks @xDaijobu I'll take a look on it

**Comment by StonedHackerman** [comment]
Camera API and a CameraView should definitelly be in the essentials! Plus in Xamarin Essentials, they already implemented it even for UWP, all that is needed is to port it here (unfortuanatelly I'm completly new to this, but it's said everywhere, that porting Xamarin to MAUI should be simple no?)

**Reply by pictos** [reply]
@StonedHackerman is kind of a different implementation, Essentials use the Platform-specific Camera app, in other words, you need to use another application to take a picture. CameraView otherwise is a control, like a Button, that you can use inside your own application so no need to use external applications to take pictures. And you also can customize the `Take Photo View`, like Instagram or TikTok do.

Well, that is not simple to port over .NET MAUI, TBH isn't easy to port from XCT to here.

**Reply by danielftz** [reply]
> @StonedHackerman is kind of a different implementation, Essentials use the Platform-specific Camera app, in other words, you need to use another application to take a picture. CameraView otherwise is a control, like a Button, that you can use inside your own application so no need to use external applications to take pictures. And you also can customize the `Take Photo View`, like Instagram or TikTok do.
> 
> Well, that is not simple to port over .NET MAUI, TBH isn't easy to port from XCT to here.

On that note, what is the status of the CameraView currently, and what can I I do to contribute to it?

**Reply by pictos** [reply]
So, we will freeze new features until GA, so I'm not working on CameraView right now. If you want to help, you can port the iOS and Windows implementations to Handler arch. and open a PR in this repo. For Android, we will change from `camera2Api` to `cameraX`, so that will need more spikes on my end before came with an implementation.

If you want to port the iOS and Windows implementation, please let me know

**Comment by CliffAgius** [comment]
Closed as CameraView is open as an Issue #1080 and being worked on.
